lib/services/user_service.ex

```diff
@@ -12,17 +12,18 @@ defmodule SlackCoder.Services.UserService do
     if db_user, do: {:ok, user}, else: {:new, user}
   end
 
-  def save(user, changeset) do
-    if changeset.changes[:slack] || changeset.changes[:github] do
-      SlackCoder.Users.Supervisor.stop_user(user)
-      user = Repo.update(changeset)
-      SlackCoder.Users.Supervisor.start_user(user)
-    else
-      SlackCoder.Users.Supervisor.start_user(user)
-    end
+  def save(%Ecto.Changeset{data: user} = changeset) do
+    pre_save(changeset.changes, user)
+    changeset
+    |> Repo.update!()
+    |> SlackCoder.Users.Supervisor.start_user()
   end
 
   defp update_user(user, db_user) do
     User.changeset(db_user || %User{}, user) |> Repo.insert_or_update!
   end
+
+  defp pre_save(%{slack: _}, user), do: SlackCoder.Users.Supervisor.stop_user(user)
+  defp pre_save(%{github: _}, user), do: SlackCoder.Users.Supervisor.stop_user(user)
```

Comment:
So not a big deal for this case just some anecdotal info... normally if it were a value you could do something like:

```elixir
defp pre_save(%{something: value}) when value in [1,2,3], do: :blah
```

But the ErlangVM won't allow you to do that with keys, so the Elixir compiler complains about it. A way around not repeating that same code would be to generate those two methods (which would write the code you have above using metaprogramming). That would look something like:

```elixir
for key <- [:slack, :github, :whatever, :else, :you, :wanted] do
  defp pre_save(%{unquote(key) => _}, user), do: SlackCoder.Users.Supervisor.stop_user(user)
end
# And then here you can do any fallback cases like
defp pre_save(_, _user), do: nil
```

For just these 2, it isnt really worth it since the complexity of adding that vs just having two single line copies of the same function call isn't worth the tradeoff. But its good to keep in mind for the future.